### PR TITLE
Metrics dedup for MLflow logger

### PR DESCRIPTION
**Pull request by chenmoneygithub**
# What does this PR do?

Context: MLflow has receiving lots of data... so we want to save some budget.

In this PR, if we detect that a metric has the same value as the previous step, we don't log it unless it meets one of the following 2 criteria:

- There has been `log_duplicated_metric_every_n_steps` skipped since last step logging, default to 100.
- There has been `log_duplicated_metric_every_n_millis` passed since last step logging, default to 600000 (10 minutes).

We put this exception for sampling purpose, otherwise it will be challenging to capture the duplicated values during sampling, despite the duplicated value occupies a significant portion of the metric history.

Tested with LLM foundry quickstart, and 3 setup: 1) no dedupping 2) log duplicated steps per 2 steps 3) log duplicated steps per 10 steps. No problem spotted. See testing result below:

image

image

image

# What issue(s) does this change relate to?

# Before submitting
- [ ] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/main/CONTRIBUTING.md)?
- [ ] Is this change a documentation change or typo fix? If so, skip the rest of this checklist.
- [ ] Was this change discussed/approved in a GitHub issue first? It is much more likely to be merged if so.
- [ ] Did you update any related docs and document your change?
- [ ] Did you update any related tests and add any new tests related to your change? (see [testing](https://github.com/mosaicml/composer/blob/main/CONTRIBUTING.md#running-tests))
- [ ] Did you run the tests locally to make sure they pass?
- [ ] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/mosaicml/composer/blob/main/CONTRIBUTING.md#prerequisites))

### CODE REVIEW

**Review by mvpatel2000 — COMMENTED** [review]
Please add manual test + link to run/plots!

File: composer/loggers/mlflow_logger.py [hunk 1]
```diff
@@ -118,6 +120,11 @@ class MLFlowLogger(LoggerDestination):
         logging_buffer_seconds (int, optional): The amount of time, in seconds, that MLflow
             waits before sending logs to the MLflow tracking server. Metrics/params/tags logged
             within this buffer time will be grouped in batches before being sent to the backend.
+        log_duplicated_metric_every_n_steps (int, optional): The number of steps to wait before
+            logging the duplicated metric value. Duplicated metric value means the new step has the
+            same value as the previous step. (default: ``100``)
+        log_duplicated_metric_every_n_millis (int, optional): The number of milliseconds to wait
+            before logging the duplicated metric value. (default: ``600000``)
```
Comment (mvpatel2000) [inline composer/loggers/mlflow_logger.py:127]: honestly this is probably fine to remove, I can't think of a good use case for this and it'll simplify things

Comment (chenmoneygithub) [reply composer/loggers/mlflow_logger.py]: sg!

**Review by mvpatel2000 — APPROVED** [review]
LGTM. But do we need to log every `n` steps if its a duplicated value? why not just never log it?

**Pull request merged**